examples/cookbook/agents/documentation_agent/agent.py

```diff
@@ -0,0 +1,134 @@
+
+
+@openai.call(
+    "gpt-4o-mini",
+    response_model=list[Relevance],
+    json_mode=True,
+)
+@prompt_template(
+    """
+    SYSTEM:
+    Document Relevance Assessment
+    Given a list of documents and a question, determine the relevance of each document to answering the question.
+    Input
+
+    A question
+    A list of documents, each with an ID and content summary
+
+    Task
+
+    Analyze each document for its relevance to the question.
+    Assign a relevance score from 1-10 for each document.
+    Provide a reason for each score.
+
+    Scoring Guidelines
+
+    Consider both direct and indirect relevance to the question.
+    Prioritize positive, affirmative information over negative statements.
+    Assess the informativeness of the content, not just keyword matches.
+    Consider the potential for a document to contribute to a complete answer.
+
+    Important Notes
+
+    Exclude documents with no relevance to the question.
+    Be cautious with negative statements - they may be relevant but are often less informative than positive ones.
+    Consider how multiple documents might work together to answer the question.
+
+    Documents:
+    {documents}
+
+    USER: 
+    query:
+    {query}
+    """
+)
+def llm_query_rerank(documents: list[dict], query: str): ...
+
+
+def get_documents(query: str) -> str:
+    """The get_documents tool that retrieves Mirascope documentation based on the
+    relevance of the query"""
+
+    query_bundle = QueryBundle(query)
+    retriever = VectorIndexRetriever(
+        index=cast(VectorStoreIndex, loaded_index),
+        similarity_top_k=10,
+    )
+    retrieved_nodes = retriever.retrieve(query_bundle)
+    choice_batch_size = 5
+    top_n = 2
+    results: list[Relevance] = []
+    for idx in range(0, len(retrieved_nodes), choice_batch_size):
+        nodes_batch = [
+            {
+                "id": idx + id,
+                "text": node.node.get_text(),  # pyright: ignore[reportAttributeAccessIssue]
+                "semantic_score": node.score,
+            }
+            for id, node in enumerate(retrieved_nodes[idx : idx + choice_batch_size])
+        ]
+        print(nodes_batch)
+        results += llm_query_rerank(nodes_batch, query)
+    results = sorted(results, key=lambda x: x.score or 0, reverse=True)[:top_n]
+
+    return "\n".join(result.document for result in results)
+
+
+class MirascopeBot(BaseModel):
+    @openai.call(
+        "gpt-4o-mini",
+    )
+    @prompt_template(
+        """
+        SYSTEM:
+        You are an AI Assistant that is an expert at answering questions about Mirascope.
+        Here is the relevant documentation to answer the question.
+
+        Context:
+        {context}
+
+        USER:
+        {question}
+        """
+    )
+    def _call(self, context: str, question: str): ...
+
+    def _step(self, question: str):
+        context = get_documents(question)
```

Comment:
nit: why get the documents here rather than inside of `_call` as a computed field?

for example:

```python
@openai.call(...)
@prompt_template(...)
def _call(self, question: str) -> openai.OpenAIDynamicConfig:
    return {"computed_fields": {"context": get_documents(question)}}
```